x/twap/api_test.go

```diff
@@ -837,3 +841,70 @@ func (s *TestSuite) TestGetArithmeticTwapToNow_ThreeAsset() {
 		})
 	}
 }
+
+func (s *TestSuite) TestGeometricTwapToNow_BalancerPool_Randomized() {
+	seed := int64(1)
+	r := rand.New(rand.NewSource(seed))
+	retries := 1000
+
+	type randTestCase struct {
+		elapsedTimeMs int
+		weightA       sdk.Int
+		tokenASupply  sdk.Int
+		weightB       sdk.Int
+		tokenBSupply  sdk.Int
+	}
+
+	maxUint64 := ^uint(0)
+
+	for i := 0; i < retries; i++ {
+		elapsedTimeMs := sdkrand.RandIntBetween(r, 1, int(maxUint64>>1))
+		weightA := sdk.NewInt(int64(sdkrand.RandIntBetween(r, 1, 1000)))
+		tokenASupply := sdk.NewInt(int64(sdkrand.RandIntBetween(r, 10_000, 1_000_000_000)))
+
+		tokenBSupply := sdk.NewInt(int64(sdkrand.RandIntBetween(r, 10_000, 1_000_000_000)))
+		weightB := sdk.NewInt(int64(sdkrand.RandIntBetween(r, 1, 1000)))
+
+		s.Run(fmt.Sprintf("elapsedTimeMs=%d, weightA=%d, tokenASupply=%d, weightB=%d, tokenBSupply=%d", elapsedTimeMs, weightA, tokenASupply, weightB, tokenBSupply), func() {
+			s.SetupTest()
+
+			ctx := s.Ctx
+			app := s.App
+
+			assets := []balancer.PoolAsset{
+				{
+					Token:  sdk.NewCoin(denom0, tokenASupply),
+					Weight: weightA,
+				},
+				{
+					Token:  sdk.NewCoin(denom1, tokenBSupply),
+					Weight: weightB,
+				},
+			}
+
+			s.PrepareCustomBalancerPool(assets, balancer.PoolParams{
+				SwapFee: sdk.ZeroDec(),
+				ExitFee: sdk.ZeroDec(),
+			})
+
+			//
+			oldTime := ctx.BlockTime().Add(1 * time.Millisecond)
+			newTime := oldTime.Add(time.Duration(elapsedTimeMs))
+
+			ctx = ctx.WithBlockTime(newTime)
+
+			spotPrice, err := app.GAMMKeeper.CalculateSpotPrice(ctx, 1, denom1, denom0)
+			s.Require().NoError(err)
+
+			twap, err := app.TwapKeeper.GetGeometricTwapToNow(ctx, 1, denom0, denom1, oldTime)
+			s.Require().NoError(err)
+
+			osmomath.ErrTolerance{
```

Comment:
Out of curiosity, why not check multiplicative tolerance too?

Reply:
Additive just gives a better idea of how many precision points there are here for all test cases. The goal of this test is just to see that twap is roughly close to spot price. So any reasonable tolerance would suffice.

Reply:
sounds good

Reply:
Hrmm, I think the guarantee we want is actually the multiplicative tolerance though. Since spot prices can differ in orders of magnitude, this may be the wrong order of magnitude for checking many of the pools.

Reply:
Yeah but the largest spot price possible in this setup is `1_000_000_000 / 10_000  =  100_000` which should not cause an extremely high additive difference.

I think I will adjust the parameters to allow for even greater spot prices and switch to multiplicative tolerance